### DISCUSSION TITLE
Add state_class - measurement to device_class - monetary

### DISCUSSION THREAD

**Discussion by andrewjswan** [home-assistant/architecture#1024] [Entity Models] [answered]
## Proposed change

Add `state_class` - `measurement` to `device_class` - `monetary`, this will remove the warning in sensors:
```
[homeassistant.components.sensor] Entity sensor.... (<class 'homeassistant.components....'>) is using state class 'measurement' which is impossible considering device class ('monetary') it is using; expected None or one of 'total';
```
That use `device_class` - `monetary`, like:
```
sensor:
  - name: ...
    unit_of_measurement: UAH
    state_class: measurement
    device_class: monetary
```
In order to monitor changes in monetary values, with their placement in Long term statistics. To track both current changes and long-term analysis, according to the documentation https://data.home-assistant.io/docs/statistics/

Sensor entities with a `measurement` (the sensor's state_class is `measurement`):
`mean` - hourly time-weighted average
`min` - hourly minimum
`max` - hourly maximum

## More info
Currency exchange, this is only one example, but it is exactly as an estimate of the value of one currency in another i.e. `1USD` is worth `30UAH`, in the end the value in `UAH` is `monetary` and at the same time is `measurement`. The ratio of one currency to another i.e. USD/UAH is no longer monetary.

It will also allow you to track the `change of prices` for gasoline, stocks, machinery or for example Intel Nuc 13 or HA Green, and everywhere it will be `monetary` and at the same time `measurement`.

It will be possible to track the maximum salary for all job offers, and many other things where state_class total cannot be applied, because it is a `measurement`.

https://github.com/home-assistant/core/pull/107113

**Comment by chertvl** [comment]
I agree! It would be very useful!

**Comment by frenck** [comment]
monetary is an accumulation, not a measurement.

Which integrations would need this? And is the device class needed to add for a specific reason?

Please note, almost all examples you gave are rates, which is not what this device is meant for. As a matter of fact, they should not be expressed in a monetary, but price per unit.

>  gasoline

That is a rate. E.g., 2.30EUR/l

> stocks, stocks, machinery or for example Intel Nuc 13 or HA Green

Also a rate. Price per one.

**Reply by TomBrien** [reply]
Coinbase would use this. Total value of currency held.

**Reply by frenck** [reply]
> Coinbase would use this. Total value of currency held.

That would be total? As in for the wallet?

**Reply by TomBrien** [reply]
It's the total in the wallet yes. Currently this is a total state class but I don't feel it's a great fit since it's very different to for example net energy which would also use total.

**Reply by frenck** [reply]
> I don't feel it's a great fit since it's very different to for example net energy which would also use total.

I'm not sure how that is different? Both are a total?

**Comment by andrewjswan** [comment]
> Please note, almost all examples you gave are rates, which is not what this device is meant for. As a matter of fact, they should not be expressed in a monetary, but price per unit.

`device_class: monetary` supports localization in GUI, if it is not specified I see only faceless UAH, and if `device_class: monetary` is specified I see normal name of currency of my country.

![image](https://github.com/home-assistant/architecture/assets/3609660/95423a21-7ef7-4bf9-aa24-d13992fe3d1f)
For sensor from post.

Maybe it is counted differently in each country, in our country it is all counted as a change in value in money (monetary).
But the same change of daily price of something is measured in money, because it is a change of price, as well as there are many other variants when `state_class: total` is not suitable and it is necessary to use `state_class: measurement`.

> That is a rate. E.g., 2.30EUR/l

In my country, such a designation is not accepted, we always indicate it as a price, ie. `2.30 Грн.` (UAH).

**Reply by frenck** [reply]
>  supports localization in GUI,

That is not a backend problem, we should not try to solve that there.

> Maybe it is counted differently in each country, in our country it is all counted as a change in value in money (monetary).

Right, but we don't build for your country-specific, we build for the world 😉

These are rates. It is a pricing per unit of something, which is not what this device class is for.

../Frenck

**Comment by andrewjswan** [comment]
> These are rates. It is a pricing per unit of something, which is not what this device class is for.

A change in daily price is not a rate, it's a `measurement`.

**Reply by frenck** [reply]
Sorry, it is not for the definition we are using.

**Comment by frenck** [comment, marked as answer]
Thank you for opening an architectural proposal. However, I'm going to respectfully decline it.

It turns out, it actually is about what we consider rates. The discussions to accept rates for monetary device class sensors have been discussed many times before and have been declined.

Therefore, I'm going to friendly decline this one as well.

.../Frenck

**Comment by andrewjswan** [comment]
> Thank you for opening an architectural proposal. However, I'm going to respectfully decline it.

You and I had a discussion a couple years ago about the unit of measurement `mmHg`, you gave the same answer that the world doesn't use it and it's not a standard, so we won't add to it. But in the end you added it, I don't know why.

What will change dramatically in Home Assistant if `state_class` - `measurement` to `device_class` - `monetary` is added?
Apart from convenience? What does this improvement violate? 
Nothing from my point of view, monetary can be either `measurement` or `total`.

**Reply by joostlek** [reply]
I mean, as far as I can see you want to translate the currency into a language currency. Why don't you take a step back and check on other ways to achieve this instead of being determined on using state classes to solve this.

**Reply by andrewjswan** [reply]
No, this is one of the side effects, I want to use `state_class - measurement` together with `device_class - monetary`, in order for the sensor to correctly hit Long Term stats like `state_class - measurement`, but still have all the properties of `device_class - monetary`.

**Reply by andrewjswan** [reply]
It is not only Total that is measured in money, but also Measurement. But that's in my world...

**Reply by frenck** [reply]
> You and I had a discussion a couple years ago about the unit of measurement mmHg, you gave the same answer that the world doesn't use it and it's not a standard, so we won't add to it. But in the end you added it, I don't know why.

I don't see how this is relevant to this proposal at all. I am also not sure why that is needed.

> What does this improvement violate?

It mixes up units. These are an expression of a sum of money, not a rate.

**Comment by andrewjswan** [comment]
Looks like I'm not the only one who needs it.

![image](https://github.com/home-assistant/architecture/assets/3609660/88bae063-536c-4049-bf46-88581f2b07a1)

**Reply by frenck** [reply]
That might be, but this ain't community a voting contest. None of these reactions came from core members.

As said, this has been evaluated and declined. As these are considered rates.

../Frenck

**Comment by andrewjswan** [comment]
> As said, this has been evaluated and declined. As these are considered rates.

Okay, last question, here are the dynamics of price changes for Intel Nuc 13, is this also a `rates` in your opinion?
Price is money (`monetary`), dynamics is `measurement`.
![image](https://github.com/home-assistant/architecture/assets/3609660/8f136476-af09-42b8-bb25-8051740cebfa)

> It mixes up units. These are an expression of a sum of money, not a rate.

Those. price is not money (`monetary`)? 
The day before yesterday the price for what I need to track was 100, yesterday 90, and today 110. This is a measurement and not a sum of money (`total`).

**Reply by frenck** [reply]
Sorry, but you make me run in circles at this point. This is considered a rate: Price per unit and not what device class is for.

**Comment by andrewjswan** [comment]
> This is considered a rate: Price per unit and not what device class is for.

English is not my native language, but the Cambridge Dictionary interprets the word `monetary` as - `pecuniary`, `moneyed`, `nummary`, `nummulary`, `material`, it also has examples of the use of this word, and judging by the context it can be both general accumulation (total) and measurement, and the same rates (strange as it may seem).
Those, adding `state_class` - `measurement` to `device_class` - `monetary` - seems quite reasonable and logical.

The Home Assistant developer documentation says:
|Constant|Supported units|Description|
|---|---|---|
| SensorDeviceClass.MONETARY | ISO 4217 | Monetary value with a currency.|

Those, any monetary value indicating the currency, price is a monetary value indicating the currency (Intel Nuc - $700). 
Price is a measurement. Because $700 is a monetary value indicating a currency, but it is not an accumulation (total).

Those the documentation is incorrect and the restriction on use is artificially introduced. 
Despite the fact that there is no confusion in units of measurement here and there will not be, the price (measurement) is money, the amount of annual income (total) is money.

**Reply by frenck** [reply]
Right, but we are talking about what this means in HA, not in the dictionary.

I'm going to halt this conversation at his point. Explanation has been given, and the proposal has been declined.

**Comment by andrewjswan** [comment]
I'll just leave this here 
https://github.com/home-assistant/core/issues/58458

**Reply by frenck** [reply]
Hiding because not related. Additionally, I've seen you've moved the selected answer, which is not acceptable in this case. You don't have to agree with our design decisions, but you'd have to accept them.

Going to lock down this discussion for that reason.
